commons/src/main/java/org/infinispan/commons/marshall/ProtoStreamMarshaller.java

```diff
@@ -68,4 +77,56 @@ protected ByteBuffer objectToBuffer(Object o, int estimatedSize) throws IOExcept
    public MediaType mediaType() {
       return MediaType.APPLICATION_PROTOSTREAM;
    }
+
+   /**
+    * A type id mapper that maps old and new type ids back and forth.
+    */
+   public static WrappedMessageTypeMapper getLegacyWrappedMessageTypeMapper() {
+      // to support the 9.x legacy, the values starting at 1000000 were used for WrappedMessage and remote query objects
+      // and need to be mapped to the new 0..65535 range
+      return new WrappedMessageTypeMapper() {
+
+         private static final int LEGACY_WRAPPED_MESSAGE = 1000000;
+         private static final int LEGACY_REMOTE_QUERY_REQUEST = 1000101;
+         private static final int LEGACY_REMOTE_QUERY_RESPONSE = 1000102;
+         private static final int LEGACY_ICKLE_FILTER_RESULT = 1000103;
+         private static final int LEGACY_ICKLE_CONTINUOUS_QUERY_RESULT = 1000104;
+
+         @Override
+         public int mapTypeId(int typeId, boolean isReading, ImmutableSerializationContext ctx) {
+            if (isReading) {
+               switch (typeId) {
+                  // historically, the values above 1000000 were used for internal stuff
+                  // mapping them unconditionally is OK
+                  case LEGACY_WRAPPED_MESSAGE:
+                     return ProtoStreamTypeIds.WRAPPED_MESSAGE;
+                  case LEGACY_REMOTE_QUERY_REQUEST:
+                     return ProtoStreamTypeIds.REMOTE_QUERY_REQUEST;
+                  case LEGACY_REMOTE_QUERY_RESPONSE:
+                     return ProtoStreamTypeIds.REMOTE_QUERY_RESPONSE;
+                  case LEGACY_ICKLE_FILTER_RESULT:
+                     return ProtoStreamTypeIds.ICKLE_FILTER_RESULT;
+                  case LEGACY_ICKLE_CONTINUOUS_QUERY_RESULT:
+                     return ProtoStreamTypeIds.ICKLE_CONTINUOUS_QUERY_RESULT;
+               }
+            } else {
+               // when writing back we map the new ids back to the legacy ones
+               // TODO [anistor] this has to happen conditionally: only if new server + old client or old server + new client, where 'new' means HR version >= 3.0
```

Comment:
I've fixed the tests. All that remains to be fixed now is this TODO,  @ryanemerson   :grin:

Reply:
The easy part! haha

Something that is still missing on the protostream side, is we need a way to toggle the use of the mapper if the ^ condition is true. Currently the a `SerializationContext` in the server is shared amongst many threads, so we can't have any state in the mapper itself.

This is what I was proposing in [IPROTO-119](IPROTO-119?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=13791385#comment-13791385).

For client compatibility at least, we just need the ability to enable/disable the mapper at runtime per protostream call.